### ISSUE TITLE
How to load pretrained EVA-02 model? And how to inference?

### ISSUE THREAD

**Issue by 1049451037** [baaivision/EVA#56]
I use the following script to load model, however failed:

```python
from asuka.modeling_pretrain import eva02_large_patch14_xattn_fusedLN_NaiveSwiGLU_subln_RoPE_xavier_normal_init
import huggingface_hub
import torch

path = huggingface_hub.hf_hub_download('Yuxin-CV/EVA-02', 'eva02_L_pt_m38m_p14.pt', subfolder='eva02/pt')
print(path)
model = eva02_large_patch14_xattn_fusedLN_NaiveSwiGLU_subln_RoPE_xavier_normal_init()
checkpoint = torch.load(path, map_location="cpu")
model.load_state_dict(checkpoint["module"])
```

It says the shape of parameters are not consistent.

```
 checkpoint, the shape in current model is torch.Size([256, 64]).
        size mismatch for blocks.22.attn.rope.freqs_sin: copying a param with shape torch.Size([16, 16, 64]) from checkpoint, the shape in current model is torch.Size([256, 64]).
        size mismatch for blocks.23.attn.rope.freqs_cos: copying a param with shape torch.Size([16, 16, 64]) from checkpoint, the shape in current model is torch.Size([256, 64]).
        size mismatch for blocks.23.attn.rope.freqs_sin: copying a param with shape torch.Size([16, 16, 64]) from checkpoint, the shape in current model is torch.Size([256, 64]).
        size mismatch for lm_head.weight: copying a param with shape torch.Size([1024, 1024]) from checkpoint, the shape in current model is torch.Size([768, 1024]).
        size mismatch for lm_head.bias: copying a param with shape torch.Size([1024]) from checkpoint, the shape in current model is torch.Size([768]).
```

How to fix this? And by the way, how to inference? What is the input format?

**Comment by Yuxin-CV** [comment]
Hi @1049451037, if you just want to load the pre-trained weights, nothing need to be fixed:

- for the size mismatch issue of rope, please refer to https://github.com/baaivision/EVA/tree/master/EVA-02/asuka#fine-tuning

- the lm_head dim depends on the teacher, for EVA-CLIP, it is 1024, the default of EVA-02-L is 768. You won't need lm_head if you just want to ft EVA-02.

**Comment by Yuxin-CV** [comment]
If you want to inference with fine-tuned EVA-02, say image classification, please follow https://github.com/baaivision/EVA/tree/master/EVA-02/asuka#fine-tuning and load the corresponding fine-tuned EVA-02 weights.

**Comment by 1049451037** [comment]
emm... However, `eva02_L_pt_m38m_p14.pt` is EVA-02-L actually. So why it is 1024?

**Comment by 1049451037** [comment]
I just want to load the pretrained eva-02 and use it as a feature extractor. Could you give me some explanation about the input shape for the model?

**Comment by 1049451037** [comment]
For my case, `lm_head` is necessary... So I have to figure out where is wrong. Is the model config or checkpoint param?

**Comment by 1049451037** [comment]
Thank you so much! I get it.

**Comment by 1049451037** [comment]
Sorry to bother you again... The rope shape mismatch cannot be ignored, because it is an error, not a warning... How to fix it?

**Comment by Yuxin-CV** [comment]
try

```python
model.load_state_dict(checkpoint["module"], strict=False)
```

**Comment by 1049451037** [comment]
`strict=False` doesn't work. It still throws shape mismatch error.

**Comment by Quan-Sun** [comment]
@1049451037 Removing all related params from state_dict
https://github.com/baaivision/EVA/blob/master/EVA-CLIP/rei/eva_clip/factory.py#L110

**Comment by Quan-Sun** [comment]
@1049451037 Removing all related params from state_dict
https://github.com/baaivision/EVA/blob/master/EVA-CLIP/rei/eva_clip/factory.py#L110

**Comment by 1049451037** [comment]
Moreover, the key point is I don't know which shape is intended. Is (16, 16, 64) or (256, 64).

**Comment by Yuxin-CV** [comment]
> Moreover, the key point is I don't know which shape is intended. Is (16, 16, 64) or (256, 64).

(256, 64)

**Comment by Yuxin-CV** [comment]
> @1049451037 Removing all related params from state_dict https://github.com/baaivision/EVA/blob/master/EVA-CLIP/rei/eva_clip/factory.py#L110

del & re-init all rope related weights is also fine, because essentially ropes are not learnable weights, they are determined & fixed.

**Comment by NAVAPOOM** [comment]
Can Eva-02 create Attention Map or Grad-Cam ?

**Comment by JunyuanDeng** [comment]
> Can Eva-02 create Attention Map or Grad-Cam ?

hello, have you figured out how to create one of them?

**Comment by fengyan-cv** [comment]
Have you solved this problem?
